### ISSUE TITLE
[HTTP/2 Client] Selector loop didn't wake up

### ISSUE THREAD

**Issue by Drdoteam** [eclipse/jetty.project#435]
Hi.
I send 10 millions messages to APNS per day by using Jetty HTTP/2 Client.

Rarely, a selector loop didn't wake up forever.

Generally, the selector loop woken up like this
```
08 19:00:01.666 DEBUG o.e.jetty.io.ManagedSelector.select(234) - Selector loop woken up from select, 0/1 selected
```
But, after print 'Selector loop waiting on select', the selector loop didn't wake up.

This is logs.
```
08 19:00:47.309 DEBUG o.e.j.client.HttpDestination.send(197) - Queued HttpRequest[POST /3/device/bbb40ea66022258b474c57fd5c1d17529ea2a7d469d3ae0041db4342d7b76954 HTTP/1.1]@2b87982e for HttpDestination[https://api.push.apple.com]4f032437,queue=524
08 19:00:47.309 DEBUG o.e.jetty.io.ssl.SslConnection.fill(534) - SslConnection@50455aa9{NOT_HANDSHAKING,eio=0/-1,di=0} -> HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,231/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,2/0,SslConnection}{io=0/0,kio=0,kro=1}] net=0 unwrap Status = BUFFER_UNDERFLOW HandshakeStatus = NOT_HANDSHAKING bytesConsumed = 0 bytesProduced = 0
08 19:00:47.310 DEBUG o.e.jetty.io.ssl.SslConnection.fill(535) - SslConnection@50455aa9{NOT_HANDSHAKING,eio=0/-1,di=0} -> HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,231/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,2/0,SslConnection}{io=0/0,kio=0,kro=1}] filled b[0]=
08 19:00:47.310 DEBUG o.e.j.http2.HTTP2Connection.produce(180) - Filled 0 bytes
08 19:00:47.310 DEBUG o.e.j.io.AbstractConnection.fillInterested(141) - fillInterested HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,232/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,3/0,SslConnection}{io=0/0,kio=0,kro=1}]
08 19:00:47.310 DEBUG o.e.jetty.io.FillInterest.register(63) - FillInterest@7f710f76{true,AC.ReadCB@3f8eb2a0{HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,3/0,SslConnection}{io=0/0,kio=0,kro=1}]}} register AC.ReadCB@3f8eb2a0{HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,3/0,SslConnection}{io=0/0,kio=0,kro=1}]}
08 19:00:47.310 DEBUG o.e.jetty.io.FillInterest.register(77) - FillInterest@7f710f76{true,AC.ReadCB@3f8eb2a0{HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,3/0,SslConnection}{io=0/0,kio=0,kro=1}]}} register AC.ReadCB@3f8eb2a0{HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,3/0,SslConnection}{io=0/0,kio=0,kro=1}]}
08 19:00:47.310 DEBUG o.e.j.io.AbstractConnection.fillInterested(141) - fillInterested SslConnection@50455aa9{NOT_HANDSHAKING,eio=-1/-1,di=-1} -> HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,-,-,3/0,SslConnection}{io=0/0,kio=0,kro=1}]
08 19:00:47.310 DEBUG o.e.jetty.io.FillInterest.register(63) - FillInterest@a68bcd0{true,AC.ReadCB@50455aa9{SslConnection@50455aa9{NOT_HANDSHAKING,eio=-1/-1,di=-1} -> HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,SslConnection}{io=0/0,kio=0,kro=1}]}} register AC.ReadCB@50455aa9{SslConnection@50455aa9{NOT_HANDSHAKING,eio=-1/-1,di=-1} -> HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,SslConnection}{io=0/0,kio=0,kro=1}]}
08 19:00:47.310 DEBUG o.e.jetty.io.FillInterest.register(77) - FillInterest@a68bcd0{true,AC.ReadCB@50455aa9{SslConnection@50455aa9{NOT_HANDSHAKING,eio=-1/-1,di=-1} -> HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,SslConnection}{io=0/0,kio=0,kro=1}]}} register AC.ReadCB@50455aa9{SslConnection@50455aa9{NOT_HANDSHAKING,eio=-1/-1,di=-1} -> HTTP2ClientConnection@3f8eb2a0[DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,SslConnection}{io=0/0,kio=0,kro=1}]}
08 19:00:47.310 DEBUG o.e.j.io.SelectChannelEndPoint.changeInterests(248) - changeInterests p=false 0->1 for SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,SslConnection}{io=0/1,kio=0,kro=1}
08 19:00:47.310 DEBUG o.e.jetty.io.ManagedSelector.submit(116) - Queued change SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,0/0,SslConnection}{io=0/1,kio=0,kro=1}:runUpdateKey on org.eclipse.jetty.io.ManagedSelector@511309d0 id=0 keys=1 selected=0
08 19:00:47.310 DEBUG o.e.j.u.t.s.ExecuteProduceConsume.produceAndRun(165) - EPR Prod/org.eclipse.jetty.http2.HTTP2Connection$HTTP2Producer@4042b703 produced null
08 19:00:47.310 DEBUG o.e.jetty.io.ManagedSelector.select(234) - Selector loop woken up from select, 0/1 selected
08 19:00:47.310 DEBUG o.e.j.u.t.s.ExecuteProduceConsume.produceAndRun(228) - EPR Idle/org.eclipse.jetty.http2.HTTP2Connection$HTTP2Producer@4042b703 produce exit
08 19:00:47.311 DEBUG o.e.jetty.io.ManagedSelector.runChange(214) - Running change SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,1/0,SslConnection}{io=0/1,kio=0,kro=1}:runUpdateKey
08 19:00:47.311 DEBUG o.e.jetty.io.ssl.SslConnection.onFillable(210) - onFillable exit DecryptedEndPoint@1c1c527{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,1/0,HTTP2ClientConnection}->SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,1/0,SslConnection}{io=0/1,kio=0,kro=1}
08 19:00:47.311 DEBUG o.e.j.io.SelectChannelEndPoint.updateKey(214) - Key interests updated 0 -> 1 on SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,1/0,SslConnection}{io=1/1,kio=1,kro=1}
08 19:00:47.311 DEBUG o.e.jetty.io.ManagedSelector.select(231) - Selector loop waiting on select
08 19:00:47.311 DEBUG o.e.j.u.t.s.ExecuteProduceConsume.produceAndRun(215) - EPR Prod/org.eclipse.jetty.io.ManagedSelector$SelectorProducer@2e14cbe1 ran SelectChannelEndPoint@61ee3a68{api.push.apple.com/17.172.234.15:443<->37915,Open,in,out,FI,-,1/0,SslConnection}{io=1/1,kio=1,kro=1}:runFillable
08 19:00:47.311 DEBUG o.e.j.u.t.s.ExecuteProduceConsume.produceAndRun(228) - EPR Prod/org.eclipse.jetty.io.ManagedSelector$SelectorProducer@2e14cbe1 produce exit
08 19:00:47.311 DEBUG o.e.j.u.t.QueuedThreadPool.run(574) - ran EPR Prod/org.eclipse.jetty.io.ManagedSelector$SelectorProducer@2e14cbe1
```

This is thread dump.
```
"pool-4-thread-933" #1005 prio=5 os_prio=0 tid=0x00007f2adc72b800 nid=0xcd6 waiting on condition [0x00007f2a682c4000]
 java.lang.Thread.State: WAITING (parking)
  at sun.misc.Unsafe.park(Native Method)
  - parking to wait for  <0x00000000c08cbc18> (a java.util.concurrent.CountDownLatch$Sync)
  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
  at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
  at org.eclipse.jetty.client.util.FutureResponseListener.get(FutureResponseListener.java:100)
  at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:653)
  at com.toast.cloud.push.agent.executor.apns.http2.ApnsHttp2Client.send(ApnsHttp2Client.java:109)
  at com.toast.cloud.push.agent.executor.apns.http2.ApnsHttp2Runnable.run(ApnsHttp2Runnable.java:129)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)
```

I suppose APNS didn't respond.
Have you any opinion ?

So... first, I set timeout and keep an eye on it.

Lastly, What are you thinking about adding default timeout ?
HttpRequest wait forever when timeout is not set.
```java
    @Override
    public ContentResponse send() throws InterruptedException, TimeoutException, ExecutionException
    {
        FutureResponseListener listener = new FutureResponseListener(this);
        send(this, listener);

        try
        {
            long timeout = getTimeout();
            if (timeout <= 0)
                return listener.get();

            return listener.get(timeout, TimeUnit.MILLISECONDS);
        }
        catch (Throwable x)
        {
            // Differently from the Future, the semantic of this method is that if
            // the send() is interrupted or times out, we abort the request.
            abort(x);
            throw x;
        }
    }
```

**Comment by sbordet** [comment]
It's a bit difficult to say something because the logs are incomplete and do not show what thread emitted what line. You should configure your logging system to print out at least the thread id.

Having said that, the log snippet looks normal to me in that the client registered for read interest. There is not much more, so I cannot tell how much time has passed since that read interest registration.

In general, it is good practice to add a total timeout to requests performed against remote servers:

```java
client.newRequest(host, port)
        .timeout(10, TimeUnit.SECONDS)
        ...
        .send();
```

I am closing this issue as it does not look there is anything wrong. If you have more information, please reopen or file a new issue.

Thanks !

**Comment by Drdoteam** [comment]
@sbordet Ok, thank you for reply!

**Comment by Drdoteam** [comment]
@sbordet Hi, timeout exception is occurred often when log level is 'DEBUG'. 
Other log level, timeout exception is not occurred... 
Is it known issue ?

**Comment by sbordet** [comment]
There are no known issues.

A `TimeoutException` is normal for a client when the server does not respond, so per se a `TimeoutException` is not indicative of any problem.

**Comment by Drdoteam** [comment]
@sbordet Thank you for reply.
I will debug about this phenomenon more.

**Comment by jeffmay** [comment]
@gimbimloki any luck finding the cause of this one?

**Comment by gimbimloki** [comment]
@jeffmay I just set timeout and handled timeout exception. 
Eventually, I couldn't find anything and concluded APNS server didn't response.
